src/main/kotlin/com/exactpro/th2/processor/core/message/protobuf/controller/CradleMessageGroupController.kt

```diff
@@ -16,33 +16,74 @@
 import com.exactpro.th2.common.grpc.EventID
 import com.exactpro.th2.common.grpc.MessageGroup
 import com.exactpro.th2.common.grpc.MessageGroupBatch
+import com.exactpro.th2.common.util.toInstant
+import com.exactpro.th2.common.utils.message.id
+import com.exactpro.th2.common.utils.message.timestamp
+import com.exactpro.th2.dataprovider.lw.grpc.EventLoadedStatistic
 import com.exactpro.th2.dataprovider.lw.grpc.MessageLoadedStatistic
 import com.exactpro.th2.processor.api.IProcessor
+import com.exactpro.th2.processor.core.Controller
+import com.exactpro.th2.processor.core.HandleMessageException
 import com.exactpro.th2.processor.core.configuration.MessageKind
-import com.exactpro.th2.processor.core.message.protobuf.controller.state.CradleMessageGroupState
-import com.exactpro.th2.processor.core.state.StateUpdater
+import com.exactpro.th2.processor.utility.ifTrue
 import com.google.protobuf.Timestamp
 
 internal class CradleMessageGroupController(
-    processor: IProcessor,
+    private val processor: IProcessor,
     intervalEventId: EventID,
     startTime: Timestamp,
     endTime: Timestamp,
     kinds: Set<MessageKind>,
     bookToGroups: Map<String, Set<String>>
-) : MessageGroupController(
-    processor,
+) : Controller<MessageGroupBatch> (
     intervalEventId
 ) {
     private val cradleMessageGroupState = CradleMessageGroupState(startTime, endTime, kinds.map(MessageKind::grpcKind).toSet(), bookToGroups)
 
     override val isStateComplete: Boolean
         get() = super.isStateComplete && cradleMessageGroupState.isStateEmpty
 
-    override fun updateActualState(func: StateUpdater<MessageGroupBatch, MessageGroup>.() -> Unit): Boolean = cradleMessageGroupState.plus(func)
+    override fun actual(batch: MessageGroupBatch) {
+        var needSignal = false
+        for (group in batch.groupsList) {
+            runCatching {
+                needSignal = needSignal or updateActualState(batch, group)
+                for (anyMessage in group.messagesList) {
+                    runCatching {
+                        updateLastProcessed(anyMessage.timestamp.toInstant())
+                        handle(anyMessage)
+                    }.onFailure { e ->
+                        throw HandleMessageException(listOf(anyMessage.id), cause = e)
+                    }
+                }
+            }.onFailure { e ->
+                throw HandleMessageException(group.messagesList.map { it.id }, cause = e)
+            }
+        }
+        needSignal.ifTrue(::signal)
+    }
+    override fun expected(loadedStatistic: MessageLoadedStatistic) {
+        updateExpectedState(loadedStatistic).ifTrue(::signal)
+        super.expected(loadedStatistic)
+    }
 
-    override fun updateExpectedState(loadedStatistic: MessageLoadedStatistic): Boolean = cradleMessageGroupState.minus(loadedStatistic)
+    override fun expected(loadedStatistic: EventLoadedStatistic) {
+        throw UnsupportedOperationException()
+    }
```

Comment:
Just in case you would want to make code a bit shorter
```suggestion
    override fun expected(loadedStatistic: EventLoadedStatistic): Unit = throw UnsupportedOperationException()
```